### ISSUE TITLE
Examine behavior dealing with writing state manifests and gc 

### ISSUE THREAD

**Issue by kellymclaughlin** [basho/riak_cs#533]
We need to verify and examine the behavior of `riak_cs` when dealing with writing state manifests when deletes or overwrites happen. There are several cases to examine and we need to verify that the behavior matches the behavior of S3. Each case is documented as a separate comment on this issue.

**Comment by kellymclaughlin** [comment]
**Case 1**

Found an behavioral difference between CS and S3. To reproduce:
1. Start a multipart upload to CS to overwrite an existing object
2. While the multipart upload is progressing, delete the object at that key that the mp upload is writing
3. Allow the mp upload to complete.

With CS the mp uploaded object is present and valid, but with S3 accessing the key returns `404 Not Found`. 

Retrieving the new upload from CS results in this message from s3cmd
`WARNING: MD5 signatures do not match: computed=a20fece9f1ecc8222410be7919ddefb6, received="de0743ce29e84f7e9e07210078acbc90"` 

I think this is just for multipart, a bug in s3cmd

**Comment by kellymclaughlin** [comment]
**Case 2**
1. Start a multipart upload to CS to overwrite an existing object
2. While the multipart upload is progressing, overwrite he object at that key that the mp upload is writing
3. Allow the mp upload to complete.

S3 (second overwrite finishes last due to larger file and later start): Shows the first overwrite while the second is uploading still. When second overwrite completes it is shown.
S3 (second overwrite finishes first due to smaller file, even though it started later): Shows the second overwrite when it finishes during upload of first and after first finishes.

Both during the final completing upload and after it finishes s3cmd returns a warning from CS
`WARNING: MD5 signatures do not match: computed=21608c9cef97d8baf5b493b72e895259, 
received="fa8552f380a3453f90f8c4c9429751ff"`

tldr: S3 uses write start time
CS: same behavior as S3

**Comment by kellymclaughlin** [comment]
**Case 3**
1. Start a normal upload of a large object to CS to overwrite an existing object
2. While the upload is progressing, delete the object at that key that the upload is writing
3. Allow the upload to complete.

CS: Returns the object
S3: Returns `404 Not Found` during and after upload

**Comment by kellymclaughlin** [comment]
**Case 4**
1. Start a normal upload of a large object to CS to overwrite an existing object
2. While the upload is progressing, overwrite the object at that key that the upload is writing
3. Allow the upload to complete.

S3 (second overwrite finishes last due to larger file and later start): Shows the first overwrite while the second is uploading still. When second overwrite completes it is shown.
S3 (second overwrite finishes first due to smaller file, even though it started later): Shows the second overwrite when it finishes during upload of first and after first finishes.

tldr: S3 uses write start time
CS: same behavior as S3

**Comment by shino** [comment]
Just for your information:

I am writing script for debugging which shows manifest states and first block existence:
https://gist.github.com/shino/3b4a3cda3dda62d52a41
One sample output is: https://gist.github.com/shino/9b60ac26a6df03e926cb

**Comment by andrewjstone** [comment]
Closing, because the fix has been merged.
